File: gommo-backend/src/main/java/br/com/gommo/modules/rh/person/exitinterview/mapper/ExitInterviewMapper.java
package br.com.gommo.modules.rh.person.exitinterview.mapper;

import java.util.ArrayList;
import java.util.Map;

import org.springframework.stereotype.Component;

import br.com.gommo.modules.rh.person.exitinterview.dto.ExitInterviewRequestDto;
import br.com.gommo.modules.rh.person.exitinterview.dto.ExitInterviewResponseDto;
import br.com.gommo.modules.rh.person.exitinterview.entity.ExitInterview;

@Component
public class ExitInterviewMapper {
    public ExitInterview toEntity(ExitInterviewRequestDto dto) {
        return ExitInterview.builder()
                .collaboratorId(dto.getCollaboratorId())
                .interviewDate(dto.getInterviewDate())
                .departureReason(dto.getDepartureReason())
                .feedback(dto.getFeedback())
                .wouldRecommend(dto.getWouldRecommend())
                .interviewStatus(dto.getInterviewStatus())
                .relationshipType(dto.getRelationshipType())
                .collaboratorName(dto.getCollaboratorName())
                .registrationNumber(dto.getRegistrationNumber())
                .jobPositionName(dto.getJobPositionName())
                .jobPositionId(dto.getJobPositionId())
                .departmentName(dto.getDepartmentName())
                .departmentId(dto.getDepartmentId())
                .companyName(dto.getCompanyName())
                .managerName(dto.getManagerName())
                .admissionOrContractStartDate(dto.getAdmissionOrContractStartDate())
                .terminationOrContractEndDate(dto.getTerminationOrContractEndDate())
                .tenureDays(dto.getTenureDays())
                .terminationType(dto.getTerminationType())
                .interviewerName(dto.getInterviewerName())
                .mainReason(dto.getMainReason())
                .secondaryReasons(dto.getSecondaryReasons() == null ? new ArrayList<>() : dto.getSecondaryReasons())
                .detailedReason(dto.getDetailedReason())
                .ratings(dto.getRatings() == null ? new ArrayList<>() : dto.getRatings())
                .openAnswers(dto.getOpenAnswers() == null ? new ArrayList<>() : dto.getOpenAnswers())
                .wouldReturn(dto.getWouldReturn())
                .companyWouldRehire(dto.getCompanyWouldRehire())
                .rehireNotes(dto.getRehireNotes())
                .returnChecklist(dto.getReturnChecklist() == null ? new ArrayList<>() : dto.getReturnChecklist())
                .templateKey(dto.getTemplateKey())
                .templateVersion(dto.getTemplateVersion())
                .templatePayload(dto.getTemplatePayload() == null ? Map.of() : dto.getTemplatePayload())
                .build();
    }

    public void updateEntity(ExitInterview entity, ExitInterviewRequestDto dto) {
        entity.setCollaboratorId(dto.getCollaboratorId());
        entity.setInterviewDate(dto.getInterviewDate());
        entity.setDepartureReason(dto.getDepartureReason());
        entity.setFeedback(dto.getFeedback());
        entity.setWouldRecommend(dto.getWouldRecommend());
        entity.setInterviewStatus(dto.getInterviewStatus());
        entity.setRelationshipType(dto.getRelationshipType());
        entity.setCollaboratorName(dto.getCollaboratorName());
        entity.setRegistrationNumber(dto.getRegistrationNumber());
        entity.setJobPositionName(dto.getJobPositionName());
        entity.setJobPositionId(dto.getJobPositionId());
        entity.setDepartmentName(dto.getDepartmentName());
        entity.setDepartmentId(dto.getDepartmentId());
        entity.setCompanyName(dto.getCompanyName());
        entity.setManagerName(dto.getManagerName());
        entity.setAdmissionOrContractStartDate(dto.getAdmissionOrContractStartDate());
        entity.setTerminationOrContractEndDate(dto.getTerminationOrContractEndDate());
        entity.setTenureDays(dto.getTenureDays());
        entity.setTerminationType(dto.getTerminationType());
        entity.setInterviewerName(dto.getInterviewerName());
        entity.setMainReason(dto.getMainReason());
        entity.setSecondaryReasons(dto.getSecondaryReasons() == null ? new ArrayList<>() : dto.getSecondaryReasons());
        entity.setDetailedReason(dto.getDetailedReason());
        entity.setRatings(dto.getRatings() == null ? new ArrayList<>() : dto.getRatings());
        entity.setOpenAnswers(dto.getOpenAnswers() == null ? new ArrayList<>() : dto.getOpenAnswers());
        entity.setWouldReturn(dto.getWouldReturn());
        entity.setCompanyWouldRehire(dto.getCompanyWouldRehire());
        entity.setRehireNotes(dto.getRehireNotes());
        entity.setReturnChecklist(dto.getReturnChecklist() == null ? new ArrayList<>() : dto.getReturnChecklist());
        entity.setTemplateKey(dto.getTemplateKey());
        entity.setTemplateVersion(dto.getTemplateVersion());
        entity.setTemplatePayload(dto.getTemplatePayload() == null ? Map.of() : dto.getTemplatePayload());
    }

    public ExitInterviewResponseDto toResponse(ExitInterview entity) {
        return ExitInterviewResponseDto.builder()
                .id(entity.getId())
                .code(entity.getCode())
                .status(entity.getStatus())
                .collaboratorId(entity.getCollaboratorId())
                .interviewDate(entity.getInterviewDate())
                .departureReason(entity.getDepartureReason())
                .feedback(entity.getFeedback())
                .wouldRecommend(entity.getWouldRecommend())
                .interviewStatus(entity.getInterviewStatus())
                .relationshipType(entity.getRelationshipType())
                .collaboratorName(entity.getCollaboratorName())
                .registrationNumber(entity.getRegistrationNumber())
                .jobPositionName(entity.getJobPositionName())
                .jobPositionId(entity.getJobPositionId())
                .departmentName(entity.getDepartmentName())
                .departmentId(entity.getDepartmentId())
                .companyName(entity.getCompanyName())
                .managerName(entity.getManagerName())
                .admissionOrContractStartDate(entity.getAdmissionOrContractStartDate())
                .terminationOrContractEndDate(entity.getTerminationOrContractEndDate())
                .tenureDays(entity.getTenureDays())
                .terminationType(entity.getTerminationType())
                .interviewerName(entity.getInterviewerName())
                .mainReason(entity.getMainReason())
                .secondaryReasons(entity.getSecondaryReasons())
                .detailedReason(entity.getDetailedReason())
                .ratings(entity.getRatings())
                .openAnswers(entity.getOpenAnswers())
                .wouldReturn(entity.getWouldReturn())
                .companyWouldRehire(entity.getCompanyWouldRehire())
                .rehireNotes(entity.getRehireNotes())
                .returnChecklist(entity.getReturnChecklist())
                .templateKey(entity.getTemplateKey())
                .templateVersion(entity.getTemplateVersion())
                .templatePayload(entity.getTemplatePayload())
                .completedAt(entity.getCompletedAt())
                .canceledAt(entity.getCanceledAt())
                .cancelReason(entity.getCancelReason())
                .createdAt(entity.getCreatedAt())
                .updatedAt(entity.getUpdatedAt())
                .build();
    }
}

File: gommo-frontend/src/modules/rh/person/exitinterview/lib/exit-interview.mapper.ts
import type {
    ExitInterview,
    ExitInterviewAnswerDto,
    ExitInterviewCreateDto,
    ExitInterviewRatingDto,
    ExitInterviewReturnChecklistItemDto,
} from "@/modules/rh/person/exitinterview/dto/exit-interview.dto";
import {
    DEFAULT_OPEN_ANSWERS,
    DEFAULT_RATINGS,
    DEFAULT_RETURN_CHECKLIST,
} from "@/modules/rh/person/exitinterview/lib/exit-interview.options";

function cloneRatings(values?: ExitInterviewRatingDto[]): ExitInterviewRatingDto[] {
    const current = new Map((values ?? []).map((item) => [item.key, item]));
    return DEFAULT_RATINGS.map((item) => ({ ...item, ...(current.get(item.key) ?? {}) }));
}

function cloneAnswers(values?: ExitInterviewAnswerDto[]): ExitInterviewAnswerDto[] {
    const current = new Map((values ?? []).map((item) => [item.key, item]));
    return DEFAULT_OPEN_ANSWERS.map((item) => ({ ...item, ...(current.get(item.key) ?? {}) }));
}

function cloneChecklist(values?: ExitInterviewReturnChecklistItemDto[]): ExitInterviewReturnChecklistItemDto[] {
    const current = new Map((values ?? []).map((item) => [item.key, item]));
    return DEFAULT_RETURN_CHECKLIST.map((item) => ({ ...item, ...(current.get(item.key) ?? {}) }));
}

export function exitinterviewToFormDto(entity: ExitInterview): ExitInterviewCreateDto {
    return {
        collaboratorId: entity.collaboratorId ?? "",
        interviewDate: entity.interviewDate?.slice(0, 10) ?? "",
        departureReason: entity.departureReason ?? "",
        feedback: entity.feedback ?? "",
        wouldRecommend: entity.wouldRecommend,
        interviewStatus: entity.interviewStatus ?? "DRAFT",
        relationshipType: entity.relationshipType ?? "CLT",
        collaboratorName: entity.collaboratorName ?? "",
        registrationNumber: entity.registrationNumber ?? "",
        jobPositionName: entity.jobPositionName ?? "",
        departmentName: entity.departmentName ?? "",
        companyName: entity.companyName ?? "",
        managerName: entity.managerName ?? "",
        admissionOrContractStartDate: entity.admissionOrContractStartDate?.slice(0, 10) ?? "",
        terminationOrContractEndDate: entity.terminationOrContractEndDate?.slice(0, 10) ?? "",
        tenureDays: entity.tenureDays,
        terminationType: entity.terminationType,
        interviewerName: entity.interviewerName ?? "",
        mainReason: entity.mainReason,
        secondaryReasons: entity.secondaryReasons ?? [],
        detailedReason: entity.detailedReason ?? entity.feedback ?? "",
        ratings: cloneRatings(entity.ratings),
        openAnswers: cloneAnswers(entity.openAnswers),
        wouldReturn: entity.wouldReturn,
        companyWouldRehire: entity.companyWouldRehire,
        rehireNotes: entity.rehireNotes ?? "",
        returnChecklist: cloneChecklist(entity.returnChecklist),
        templateKey: entity.templateKey ?? "standard_exit_interview",
        templateVersion: entity.templateVersion ?? 1,
        templatePayload: entity.templatePayload ?? {},
    };
}

export const emptyExitInterviewForm = (): ExitInterviewCreateDto => ({
    collaboratorId: "",
    interviewDate: new Date().toISOString().slice(0, 10),
    departureReason: "",
    feedback: "",
    interviewStatus: "DRAFT",
    relationshipType: "CLT",
    collaboratorName: "",
    registrationNumber: "",
    jobPositionName: "",
    departmentName: "",
    companyName: "",
    managerName: "",
    admissionOrContractStartDate: "",
    terminationOrContractEndDate: "",
    interviewerName: "",
    secondaryReasons: [],
    detailedReason: "",
    ratings: cloneRatings(),
    openAnswers: cloneAnswers(),
    rehireNotes: "",
    returnChecklist: cloneChecklist(),
    templateKey: "standard_exit_interview",
    templateVersion: 1,
    templatePayload: {},
});